modulus/datapipes/gnn/lagrangian_dataset.py

```diff
@@ -0,0 +1,346 @@
+        keepdim=True,
+    )
+    # direction = displacement / distance
+    distance = torch.exp(-(distance**2) / radius**2)
+    graph.edata["x"] = torch.cat((displacement, distance), dim=-1)
+    return
+
+
+def graph_update(graph, radius):
+    # remove all previous edges and re-construct the graph using pair-wise distance
+    # TODO: use more efficient graph construction method
+    num_edges = graph.num_edges()
+    if num_edges > 0:
+        graph.remove_edges(torch.arange(num_edges, device=graph.device))
+    mesh_pos = graph.ndata["mesh_pos"]
+    edge_index = compute_edge_index(mesh_pos, radius)
+    graph.add_edges(edge_index[0], edge_index[1])
+    compute_edge_attr(graph)
+    return graph
+
+
+class LagrangianDataset(DGLDataset):
+    """In-memory MeshGraphNet Dataset for lagrangian mesh
+    Notes:
+        - This dataset prepares and processes the data available in MeshGraphNet's repo:
+            https://github.com/google-deepmind/deepmind-research/tree/master/learning_to_simulate
+
+    Parameters
+    ----------
+    name : str, optional
+        Name of the dataset, by default "dataset"
+    data_dir : _type_, optional
+        Specifying the directory that stores the raw data in .TFRecord format., by default None
+    split : str, optional
+        Dataset split ["train", "eval", "test"], by default "train"
+    num_samples : int, optional
+        Number of samples, by default 1000
+    num_steps : int, optional
+        Number of time steps in each sample, by default 600
+    noise_std : float, optional
+        The standard deviation of the noise added to the "train" split, by default 0.02
+    force_reload : bool, optional
+        force reload, by default False
+    verbose : bool, optional
+        verbose, by default False
+    """
+
+    def __init__(
+        self,
+        name="dataset",
+        data_dir=None,
+        split="train",
+        num_samples=1000,
+        num_history=5,
+        num_steps=600,
+        noise_std=0.0003,
+        radius=0.015,
+        dt=0.0025,
+        bound=[0.1, 0.9],
+        force_reload=False,
+        verbose=False,
+    ):
+        super().__init__(
+            name=name,
+            force_reload=force_reload,
+            verbose=verbose,
+        )
+        self.data_dir = data_dir
+        self.split = split
+        self.num_samples = num_samples
+        self.num_history = num_history
+        self.num_steps = num_steps
+        self.noise_std = noise_std
+        self.length = num_samples * (num_steps - 1)
+
+        path_metadata = os.path.join(data_dir, "metadata.json")
+        with open(path_metadata, "r") as file:
+            metadata = json.load(file)
+            self.dt = metadata["dt"]
+            self.radius = metadata["default_connectivity_radius"]
+            self.bound = metadata["bounds"][0]
+            self.dim = metadata["dim"]
+            self.vel_mean = torch.tensor(metadata["vel_mean"]).reshape(1, 2)
```

Comment:
@zongyi-li Use reshape(1, self.dim) to make it work with both 2D and 3D datasets.

Reply:
Thanks! fixed